timedf/benchmark_utils.py

```diff
@@ -160,56 +160,106 @@ def memory_usage():
-    max_mem = 0
-    try:
-        with open(f"/proc/{proc.pid}/status", "r") as stat:
-            for match in re.finditer(_VM_PEAK_PATTERN, stat.read()):
-                max_mem = float(match.group(1))
-                # MB conversion
-                max_mem = int(max_mem / 1024)
-                break
-    except FileNotFoundError:
-        if platform.system() == "Linux":
-            warnings.warn(f"Couldn't open `/proc/{proc.pid}/status` file. Is the process alive?")
-        else:
-            warnings.warn("Couldn't get the max memory usage on a non-Linux platform.")
-        return 0
+    def start(self):
+        """
+        Starts tracking memory in a child process.
+        """
+        self.child = multiprocessing.Process(target=self.track_in_child)
+        self.child.start()
+        self.tracking_in_progress = True
+
+    def track_in_child(self):
+        """
+        Function to track memory periodically
+
+        Tracks memory periodically, when tracking ends peak memory put to data queue.
+        """
+        max_memory_system = 0
+        while not self.stop_event.is_set():
+            time.sleep(0.001)
+            meminfo_values = self._read_meminfo()
+            current_max_memory_system = self._calculate_used_memory(meminfo_values)
+            max_memory_system = max(current_max_memory_system, max_memory_system)
+            # Htop would show value of (current_max_memory_system / 1024)Gb in Memory bar.
+        self.data_queue.put(max_memory_system)
+
+    @staticmethod
+    def _read_meminfo():
+        """
+        Read contents of /proc/meminfo.
+        """
+        meminfo_values = {}
+        with open("/proc/meminfo") as meminfo_file:
+            for line in meminfo_file:
+                key, value = line.split(":")
+                key = key.strip()
+                value = int(value.split()[0])  # Extract the numeric value and convert to int
+                meminfo_values[key] = value
+        return meminfo_values
+
+    @staticmethod
+    def _calculate_used_memory(meminfo_values):
+        """
+        Calculate used memory with the logic used in htop
+        https://github.com/htop-dev/htop/blob/main/linux/LinuxMachine.c
+
+        Parameters
+        ----------
+        meminfo_values : dict
+            Content of /proc/meminfo
+
+        Returns
+        -------
+        float
+            Calculated used memory.
+        """
+        total_mem = meminfo_values.get("MemTotal", 0)
+        cached_mem = meminfo_values.get("Cached", 0)
+        sreclaimable_mem = meminfo_values.get("SReclaimable", 0)
+        free_mem = meminfo_values.get("MemFree", 0)
+        buffers_mem = meminfo_values.get("Buffers", 0)
+        used_diff = free_mem + cached_mem + sreclaimable_mem + buffers_mem
+        usedMem = total_mem - used_diff if total_mem >= used_diff else total_mem - free_mem
+        used_mem_mb = usedMem / (1024)
+        return used_mem_mb
+
+    def get_result(self):
```

Comment:
```suggestion
    def get_memory_used(self):
```
Should we rename it?

Reply:
Yes renaming seems better.